### PR TITLE
Add Distinct operator

### LINKED ISSUE #65: Operator: Distinct
[Link to Rx.NET implementation](https://github.com/dotnet/reactive/blob/main/AsyncRx.NET/System.Reactive.Async.Linq/System/Reactive/Linq/Operators/Distinct.cs)

### PR DESCRIPTION
**Pull request by ronnygunawan**
Resolves #65

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/reaqtive/reaqtor?pullRequest=99) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/reaqtive/reaqtor?pullRequest=99) before we can accept your contribution.<br/><br/>:x: ronnygunawan [sign now](https://cla.dotnetfoundation.org/reaqtive/reaqtor?pullRequest=99)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/reaqtive/reaqtor?pullRequest=99) it.</sub>

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            private HashSet<TKey> _keySet;
+
+            public _(Distinct<TResult, TKey> parent, IObserver<TResult> observer)
+                : base(parent, observer)
+            {
+            }
+
+            protected override void OnStart()
+            {
+                if (_keySet == null)
```
Comment (bartdesmet) [inline Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs:44]: We can use `??=` null-coalescing assignment here.

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+            public void OnError(Exception error)
+            {
+                Output.OnError(error);
+                Dispose();
+            }
+
+            public void OnNext(TResult value)
+            {
+                TKey key;
```
Comment (bartdesmet) [inline Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs:68]: This local can be moved inside the `try` block.

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+                    Output.OnNext(value);
+                }
+            }
+
+            protected override void LoadStateCore(IOperatorStateReader reader)
+            {
+                base.LoadStateCore(reader);
+
+                _keySet = reader.Read<HashSet<TKey>>();
```
Comment (bartdesmet) [inline Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs:93]: This is okay for the time being, but some future work is needed to integrate https://github.com/reaqtive/reaqtor/tree/main/Reaqtor/Pearls/OperatorLocalStorage in the core product. One of the reasons operators like `Distinct` were omitted from the initial operator set was the potentially unbounded state storage, and the fact that any little change to the state of the operator causes it to re-persist (rather than persisting deltas).

We can consider whether we want to tackle this for a 1.0 release, or later. Doing it early would be beneficial to drive out rough edges of the operator local storage feature, but it's a significant piece of work. On the other hand, doing it later results in having to deal with state migration (which is already the case for other operators like `GroupBy` which can have big storage requirements).

One thing worth considering in the meantime is to support an upper bound setting, as we've done in other places (and used in production to prevent runaway state), e.g. https://github.com/reaqtive/reaqtor/blob/main/Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/GroupBy.cs#L86. This provides a temporary middle ground where users can leverage the Reaqtive operator library as an Rx variant (and as such, we can continue to complete the operator set over here, because Reaqtive is usable by itself as well, outside the bigger Reaqtor domain), while providing a dial for those hosting it in an environment where state sizes can become problematic.

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Subscribable.cs [hunk 4]
```diff
@@ -630,6 +631,68 @@ public static ISubscribable<TSource> Finally<TSource>(this ISubscribable<TSource
...
+        public static ISubscribable<TSource> Distinct<TSource>(this ISubscribable<TSource> source)
+        {
+            if (source == null)
+                throw new ArgumentNullException(nameof(source));
+
+            return new Distinct<TSource, TSource>(source, x => x, EqualityComparer<TSource>.Default);
+        }
+
+        /// <summary>
+        /// Propagates distinct elements from the source sequence using the default comparer for elements.
```
Comment (bartdesmet) [inline Reaqtive/Core/Reaqtive.Linq/Reaqtive/Subscribable.cs:650]: Use the `specified comparer` rather than `default comparer`.

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Subscribable.cs [hunk 5]
```diff
@@ -630,6 +631,68 @@ public static ISubscribable<TSource> Finally<TSource>(this ISubscribable<TSource
...
+        /// <param name="source">Source sequence whose elements to propagate.</param>
+        /// <param name="keySelector">Selector function to apply to each element to obtain a key used for equality comparison between elements.</param>
+        /// <returns>Observable sequence containing distinct elements from the source sequence.</returns>
+        public static ISubscribable<TSource> Distinct<TSource, TKey>(this ISubscribable<TSource> source, Func<TSource, TKey> keySelector)
+        {
+            return Distinct(source, keySelector, EqualityComparer<TKey>.Default);
+        }
+
+        /// <summary>
+        /// Propagates distinct elements from the source sequence using the default comparer for elements.
```
Comment (bartdesmet) [inline Reaqtive/Core/Reaqtive.Linq/Reaqtive/Subscribable.cs:675]: Same as earlier.

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Subscribable.cs [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (bartdesmet) [inline Reaqtive/Core/Reaqtive.Linq/Reaqtive/Subscribable.cs:7]: Accidentally added? I don't see it being used.

File: Reaqtor/Core/Expressions/Reaqtor.Expressions.Core/Reaqtor/ReactiveQbservable.cs [hunk 7]
```diff
@@ -502,6 +502,60 @@ public static IReactiveQbservable<TSource> Finally<TSource>(this IReactiveQbserv
...
+
+        /// <summary>
+        /// Propagates distinct elements from the source sequence using the default comparer for elements.
+        /// </summary>
+        /// <typeparam name="TSource">Type of the elements in the source sequence.</typeparam>
+        /// <param name="source">Source sequence whose elements to propagate.</param>
+        /// <returns>Observable sequence propagating the source sequence's distinct elements.</returns>
+        public static IReactiveQbservable<TSource> Distinct<TSource>(this IReactiveQbservable<TSource> source)
+        {
+            throw new NotImplementedException();
```
Comment (bartdesmet) [inline Reaqtor/Core/Expressions/Reaqtor.Expressions.Core/Reaqtor/ReactiveQbservable.cs:515]: I'd welcome a separate PR (to avoid unrelated stylistics changes here) that changes all of these to use the `=>` expression bodied syntax for all of these `throw` statements.

**Review by bartdesmet — CHANGES_REQUESTED** [review]
Thanks very much for your contribution! I've left a few notes on the north star goal of moving towards "operator local storage" in the future, but that would be a separate effort. In the meantime, it may be worth reading up on that.

File: Reaqtor/Samples/Shebang/Reaqtor.Shebang.Service/Linq/QueryOperators.cs [hunk 8]
```diff
@@ -57,6 +57,10 @@ public static async Task DefineAsync(ReactiveClientContext ctx, CancellationToke
             await ctx.DefineObservableAsync(new Uri("rx://observable/count/predicate"), (IAsyncReactiveQbservable<TSource> source, Func<TSource, bool> predicate) => Subscribable.Count<TSource>(source.AsSubscribable(), predicate).AsAsyncQbservable(), null, token).ConfigureAwait(false);
             await ctx.DefineObservableAsync(new Uri("rx://observable/delaysubscription/absolute"), (IAsyncReactiveQbservable<TSource> source, DateTimeOffset dueTime) => Subscribable.DelaySubscription<TSource>(source.AsSubscribable(), dueTime).AsAsyncQbservable(), null, token).ConfigureAwait(false);
             await ctx.DefineObservableAsync(new Uri("rx://observable/delaysubscription/relative"), (IAsyncReactiveQbservable<TSource> source, TimeSpan dueTime) => Subscribable.DelaySubscription<TSource>(source.AsSubscribable(), dueTime).AsAsyncQbservable(), null, token).ConfigureAwait(false);
+            await ctx.DefineObservableAsync(new Uri("rx://observable/distinct/"), (IAsyncReactiveQbservable<TSource> source) => Subscribable.Distinct<TSource>(source.AsSubscribable()).AsAsyncQbservable(), null, token).ConfigureAwait(false);
```
Comment (bartdesmet) [inline Reaqtor/Samples/Shebang/Reaqtor.Shebang.Service/Linq/QueryOperators.cs:60]: Also add the matching `Undefine` operations below.

**Review by bartdesmet — CHANGES_REQUESTED** [review]

File: Reaqtor/Samples/Remoting/Tests.Reaqtor.Remoting.Glitching/Operators/Distinct.Declarative.cs [hunk 9]
```diff
@@ -0,0 +1,238 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (bartdesmet) [inline Reaqtor/Samples/Remoting/Tests.Reaqtor.Remoting.Glitching/Operators/Distinct.Declarative.cs:1]: Something's a little off here. I'd expect to see this file under `Tests.Reaqtive.Linq\Operators` instead, and then have a link in the `csproj` for `Tests.Reaqtor.Remoting.Glitching`, see https://github.com/reaqtive/reaqtor/blob/main/Reaqtor/Samples/Remoting/Tests.Reaqtor.Remoting.Glitching/Tests.Reaqtor.Remoting.Glitching.csproj.

**Review by bartdesmet — CHANGES_REQUESTED** [review]

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs (hunk 3 above)
Comment (ronnygunawan) [reply Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs:102]: Should we add an analyzer that gives warning to users who try to use Reaqtor v1 operators which contain exponentially growing state?

File: Reaqtor/Core/Expressions/Reaqtor.Expressions.Core/Reaqtor/ReactiveQbservable.cs (hunk 7 above)
Comment (ronnygunawan) [reply Reaqtor/Core/Expressions/Reaqtor.Expressions.Core/Reaqtor/ReactiveQbservable.cs:515]: Sure

File: Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            }
+
+            public void OnNext(TResult value)
+            {
+                var added = false;
+                try
+                {
+                    TKey key = Params._keySelector(value);
+
+                    // TODO: Should this operator silently stop propagating when it reaches max capacity? or should it throw exception?
```
Comment (bartdesmet) [inline Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs:80]: We should propagate an exception down the `OnError` channel, just like we do for other operators that have such a threshold. Otherwise, we end up silently producing incorrect sequences, which is semantically incorrect and unacceptable.

While erroring out is not ideal (cf. my discussion on operator local storage as the way out of this), it prevents state explosion in cases where it matters (i.e. places where users opt in). Also note that these settings come from the operator context, which can be controlled on a per-subscription basis.

Concrete deployments of Reaqtor have used this as a quota mechanism where different queries get different storage quota, while others chose other policies such as conservative global settings for entire engines, or simply don't deploy operators that have potentially unbounded storage requirements. Initially, we were leaning towards the latter by simply omitting such operators (e.g. `Buffer`, `Zip`, etc.) but as legitimate use cases grew (often on bounded sequences produced by e.g. a `Window` operator), we ended up adding more of these operators (with configurable quota settings) to the Reaqtive library.

One more piece of history is worth pointing out here. When we started this project, Reaqtive was an implementation detail of Reaqtor, and was only used in that context. As we shrunk down the footprint of engines and improved the experience on manually using the Reaqtive layer (with convenience subscription visitors used to initialize, load/save state, etc.), we decided to decouple of operator library from the engine more and more. As part of the OSS effort, we truly split off the Reaqtive libraries at their own entity in the "triad" of Nuqleon, Reaqtive, Reaqtor. (There was some complexity around this especially for higher-order operators, which bank on some facilities provided by the hosting environment.) This will allow for broader reuse.

Finally, note that the setting is optional and we'll default to `int.MaxValue`, which corresponds to the Rx default (i.e. there is no bound over there).

Comment (bartdesmet) [reply Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs:80]: See https://github.com/reaqtive/reaqtor/blob/main/Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Buffer.cs#L42 for the approach we take.

File: Reaqtor/Core/Reactive/Tests.Reaqtor.Reactive.Linq.Hosted/Distinct.cs [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+
+                var res = client.Start(ctx, () =>
+                    xs.Distinct(),
+                    100, 200, 1000);
+
+                res.Messages.AssertEqual(
+                    OnNext(210, 2),
+                    OnNext(220, 3),
+                    OnNext(230, 4),
+                    OnCompleted<int>(250)
```
Comment (bartdesmet) [inline Reaqtor/Core/Reactive/Tests.Reaqtor.Reactive.Linq.Hosted/Distinct.cs:315]: We'll expect an error here, similar to https://github.com/reaqtive/reaqtor/blob/main/Reaqtive/Core/Tests.Reaqtive.Linq/Operators/Buffer.cs#L98.

**Review by bartdesmet — APPROVED** [review]
